### PR TITLE
property test multiple deposit

### PR DESCRIPTION
**Pull request by grigorevPavel**

### CODE REVIEW

File: test/integration/multiple_deposit.test.ts [hunk 1]
```diff
@@ -1,20 +1,29 @@
...
+import { ERC20RootVaultGovernance } from "../types";
+import { deposit } from "../../tasks/vaults";
+import { Address } from "hardhat-deploy/dist/types";
+import { randomBytes } from "crypto";
 
 type CustomContext = {
     erc20Vault: ERC20Vault;
     yearnVault: YearnVault;
+    erc20RootVaultNft: number;
+    usdcSupply: BigNumber;
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:23]: rename

File: test/integration/multiple_deposit.test.ts [hunk 2]
```diff
@@ -125,169 +151,273 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
-                    .connect(this.deployer)
-                    .deposit(
-                        [
-                            BigNumber.from(amountUSDC).div(numDeposits),
-                            BigNumber.from(amountWETH).div(numDeposits),
-                        ],
-                        0
+        pit(
+            `
+            multiple assymetric deposit + multiple assymetric withdraw with zero fees\n
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:156]: When fees are 0, sum of deposit[i] = sum of withdraw[j]

File: test/integration/multiple_deposit.test.ts [hunk 3]
```diff
@@ -125,169 +151,273 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+            sum of deposit[i] = sum of withdraw[j]
+        `,
+            { numRuns: RUNS.verylow, endOnFailure: true },
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:159]: Runs.mid

File: test/integration/multiple_deposit.test.ts [hunk 4]
```diff
@@ -125,169 +151,273 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+            integer({ min: 1, max: 10 }),
+            integer({ min: 1, max: 10 }),
+            integer({ min: 100_000, max: 1_000_000 }).map((x) =>
+                BigNumber.from(x.toString())
+            ),
+            integer({ min: 10 ** 11, max: 10 ** 15 }).map((x) =>
+                BigNumber.from(x.toString())
+            ),
+            async (
+                delay: number,
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:170]: Not needed

File: test/integration/multiple_deposit.test.ts [hunk 5]
```diff
@@ -125,169 +151,273 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                        .deposit(
+                            [
+                                BigNumber.from(amountUSDC).div(numDeposits),
+                                BigNumber.from(amountWETH).div(numDeposits),
+                            ],
+                            0
+                        );
+                }
+
+                const lpTokensAmount = await this.subject.balanceOf(
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:188]: Add equal lp tokens

**Review by alexeuler — CHANGES_REQUESTED** [review]

File: test/integration/multiple_deposit.test.ts [hunk 6]
```diff
@@ -97,15 +106,22 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                        .mul(5);
+                    this.usdcDeveloperSupply = BigNumber.from(10)
+                        .pow(20)
+                        .mul(5);
+
                     await mint(
                         "USDC",
                         this.deployer.address,
-                        BigNumber.from(10).pow(20).mul(5)
+                        this.usdcDeveloperSupply
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:119]: deployer

File: test/integration/multiple_deposit.test.ts [hunk 7]
```diff
@@ -125,169 +163,280 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                BigNumber.from(x.toString())
+            ),
+            async (
+                numDeposits: number,
+                numWithdraws: number,
+                amountUSDC: BigNumber,
+                amountWETH: BigNumber
+            ) => {
+                let lpAmounts: BigNumber[] = [];
+                expect(
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:186]: assert

File: test/integration/multiple_deposit.test.ts [hunk 8]
```diff
@@ -125,169 +163,280 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                        BigNumber.from(lpTokensAmount).mod(numWithdraws),
+                        [0, 0]
+                    );
+                }
+                expect(
+                    await this.subject.balanceOf(this.deployer.address)
+                ).to.deep.equals(BigNumber.from(0));
+
+                expect(
+                    await this.weth.balanceOf(this.deployer.address)
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:265]: Check +1

File: test/integration/multiple_deposit.test.ts [hunk 9]
```diff
@@ -125,169 +163,280 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                    );
+                }
+                if ((await this.subject.balanceOf(protocolTreasury)).gt(0)) {
+                    await this.subject.withdraw(
+                        protocolTreasury,
+                        BigNumber.from(2).pow(256).sub(1),
+                        [0, 0]
+                    );
+                }
+
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:414]: add expectations for fees

Comment (alexeuler) [reply test/integration/multiple_deposit.test.ts:414]: No writeoff

**Review by alexeuler — CHANGES_REQUESTED** [review]

File: contracts/vaults/ERC20RootVault.sol [hunk 10]
```diff
@@ -289,13 +288,14 @@ contract ERC20RootVault is IERC20RootVault, ERC20Token, ReentrancyGuard, Aggrega
         );
     }
 
+    //FIX
```
Comment (alexeuler) [inline contracts/vaults/ERC20RootVault.sol:291]: Why this change?

File: contracts/vaults/ERC20RootVault.sol [hunk 11]
```diff
@@ -128,6 +128,9 @@ contract ERC20RootVault is IERC20RootVault, ERC20Token, ReentrancyGuard, Aggrega
         address[] memory tokens = _vaultTokens;
         uint256[] memory tokenAmounts = new uint256[](_vaultTokens.length);
         (uint256[] memory minTvl, ) = tvl();
+        if (lpTokenAmount > balanceOf[to]) {
```
Comment (alexeuler) [inline contracts/vaults/ERC20RootVault.sol:131]: This needs to be added to some "fixes" branch for auditors

File: test/integration/multiple_deposit.test.ts [hunk 12]
```diff
@@ -125,169 +178,533 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                amountUSDC: BigNumber,
+                amountWETH: BigNumber
+            ) => {
+                let lpAmounts: BigNumber[] = [];
+                assert(
+                    (await this.subject.balanceOf(this.deployer.address)).eq(
+                        BigNumber.from(0)
+                    )
+                );
+                for (var i = 0; i < numDeposits; ++i) {
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:206]: var is deprecated, use let

File: test/integration/multiple_deposit.test.ts [hunk 13]
```diff
@@ -116,6 +126,49 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                    await sleep(this.governanceDelay);
+                    await this.erc20RootVaultGovernance
+                        .connect(this.admin)
+                        .commitDelayedStrategyParams(this.erc20RootVaultNft);
+                    const params = {
+                        forceAllowMask: ALLOW_MASK,
+                        maxTokensPerVault: 10,
+                        governanceDelay: 86400,
+                        protocolTreasury,
+                        withdrawLimit: BigNumber.from(10).pow(20),
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:156]: Make in fixture

File: test/integration/multiple_deposit.test.ts [hunk 14]
```diff
@@ -116,6 +126,49 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                    this.strategyPerformanceTreasury = randomAddress();
+
+                    await erc20RootVaultGovernance
+                        .connect(this.admin)
+                        .stageDelayedStrategyParams(this.erc20RootVaultNft, {
+                            strategyTreasury: this.strategyTreasury,
+                            strategyPerformanceTreasury:
+                                this.strategyPerformanceTreasury,
+                            privateVault: true,
+                            managementFee: 0,
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:144]: Make in fixture

File: test/integration/multiple_deposit.test.ts [hunk 15]
```diff
@@ -116,6 +126,49 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                        .connect(this.admin)
+                        .stageParams(params);
+                    await sleep(this.governanceDelay);
+                    await this.protocolGovernance
+                        .connect(this.admin)
+                        .commitParams();
+                    this.mellowOracle = await ethers.getContract(
+                        "MellowOracle"
+                    );
+                    assert(
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:168]: remove

File: test/integration/multiple_deposit.test.ts [hunk 16]
```diff
@@ -125,169 +178,533 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                const lpTokensAmount = await this.subject.balanceOf(
+                    this.deployer.address
+                );
+                expect(lpTokensAmount).to.not.deep.equals(BigNumber.from(0));
+
+                let erc20_tvl = await this.erc20Vault.tvl();
+                let yearn_tvl = await this.yearnVault.tvl();
+                let root_tvl = await this.subject.tvl();
+
+                expect(erc20_tvl[0][0].add(yearn_tvl[0][0])).to.deep.equals(
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:236]: add [1][..]

File: test/integration/multiple_deposit.test.ts [hunk 17]
```diff
@@ -125,169 +178,533 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+
+                for (let i = 0; i < numWithdraws; ++i) {
+                    await this.subject.withdraw(
+                        this.deployer.address,
+                        BigNumber.from(lpTokensAmount).div(numWithdraws),
+                        [0, 0]
+                    );
+                }
+
+                if (
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:251]: // dust
assert (lp balance < numWithdraws ** 2)
withdraw (lp balance)

File: test/integration/multiple_deposit.test.ts [hunk 18]
```diff
@@ -125,169 +178,533 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
             await this.deploymentFixture();
         });
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:180]: describe("properties") {
   const setFeesFixture = ...
}

File: test/integration/multiple_deposit.test.ts [hunk 19]
```diff
@@ -125,169 +178,533 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                    managementFee: BigNumber.from(20000000),
+                    performanceFee: BigNumber.from(200000000),
+                });
+            await sleep(this.governanceDelay);
+            await this.erc20RootVaultGovernance
+                .connect(this.admin)
+                .commitDelayedStrategyParams(this.erc20RootVaultNft);
+        });
+
+        const calculateExcepctedPerformanceFees = (
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:300]: typo

File: test/integration/multiple_deposit.test.ts [hunk 20]
```diff
@@ -125,169 +178,533 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+            integer({ min: 0, max: 5 * 86400 }),
+            integer({ min: 2, max: 10 }),
+            integer({ min: 2, max: 10 }),
+            integer({ min: 0, max: 10 }),
+            integer({ min: 0, max: 10 }),
+            async (
+                delay: number,
+                numDeposits: number,
+                numWithdraws: number,
+                ratioUSDC: number,
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:334]: add comment

Comment (alexeuler) [reply test/integration/multiple_deposit.test.ts:334]: make random float

File: test/integration/multiple_deposit.test.ts [hunk 21]
```diff
@@ -125,169 +178,533 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                    expect(
+                        await this.usdc.balanceOf(this.deployer.address)
+                    ).to.be.equal(this.usdcDeployerSupply);
+                }
+
+                let erc20_tvl = await this.erc20Vault.tvl();
+                let yearn_tvl = await this.yearnVault.tvl();
+                let root_tvl = await this.subject.tvl();
+
+                expect(erc20_tvl[0][0].add(yearn_tvl[0][0])).to.deep.equals(
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:445]: tvl[1]

File: test/integration/multiple_deposit.test.ts [hunk 22]
```diff
@@ -125,169 +143,648 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                        await this.protocolGovernance.protocolTreasury();
+
+                    const lpTokensAmount = await this.subject.balanceOf(
+                        this.deployer.address
+                    );
+                    expect(lpTokensAmount).to.not.deep.equals(
+                        BigNumber.from(0)
+                    );
+
+                    if (roundedTokensDepositRatio.eq(10 ** 3)) {
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:477]: This looks very strange

File: test/integration/multiple_deposit.test.ts [hunk 23]
```diff
@@ -125,169 +143,648 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                    let averagePrice = BigNumber.from(0);
+                    for (let i = 0; i < pricesX96.length; ++i) {
+                        averagePrice = averagePrice.add(pricesX96[i]);
+                    }
+                    averagePrice = averagePrice.div(pricesX96.length);
+
+                    let tvls = await this.subject.tvl();
+                    let minTvl = tvls[0];
+
+                    let withdrawAmounts: BigNumber[] = [];
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:590]: This all needs to be broken down into smaller pieces. It's impossible to read 500 lines of code with no sections / functions breakdown

File: test/integration/multiple_deposit.test.ts [hunk 24]
```diff
@@ -125,169 +143,793 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+
+                    let tokensDepositRatioEqualsOne: boolean =
+                        roundedTokensDepositRatio.eq(10 ** 3);
+                    let tokensDepositRatioEqualsZero: boolean =
+                        roundedTokensDepositRatio.eq(0);
+
+                    let usdcDepositAmounts: BigNumber[] = [];
+                    let wethDepositAmounts: BigNumber[] = [];
+
+                    /*
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:398]: Nice! 👍

File: test/integration/multiple_deposit.test.ts [hunk 25]
```diff
@@ -125,169 +143,793 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+                            .abs();
+
+                        /* 
+                            MANAGEMENT FEES
+                            dif <= 0.01 * 1% * fee
+                            dif * 100 * 100 <= fee
+                            dif * (10 ** 4) - fee <= 0
+                        */
+                        expect(
+                            managementFeeAbsDifference
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:832]: The abs difference should be small enough. Can we convert it to floats?

File: test/integration/multiple_deposit.test.ts [hunk 26]
```diff
@@ -125,169 +143,793 @@ contract<ERC20RootVault, DeployOptions, CustomContext>(
...
+
+                    // collect protocol fees
+                    if (
+                        (await this.subject.balanceOf(protocolTreasury)).gt(0)
+                    ) {
+                        await this.subject.withdraw(
+                            protocolTreasury,
+                            BigNumber.from(2).pow(256).sub(1),
+                            [0, 0]
+                        );
```
Comment (alexeuler) [inline test/integration/multiple_deposit.test.ts:882]: Can we also add a post-withdraw check in terms of real tokens? E.g. expect ETH == ETH

File: test/integration/multiple_deposit.test.ts (hunk 25 above)
Comment (grigorevPavel) [reply test/integration/multiple_deposit.test.ts:832]: IMO no need to use floats here because it there is a BigNumber in calculations => we must use x * 10**6 / 10**6

Right now we do not use rounded values, which IMO is better

File: test/integration/multiple_deposit.test.ts (hunk 26 above)
Comment (grigorevPavel) [reply test/integration/multiple_deposit.test.ts:882]: I already have these post-withdraw checks for WETH and USDC in the end if the test

**Review by alexeuler — APPROVED** [review]
